
1. Start with installing and loading the Biostrings library by typing the
following commands


```R
source("http://bioconductor.org/biocLite.R")
biocLite("Biostrings")
library(Biostrings)
```

    Bioconductor version 3.7 (BiocInstaller 1.30.0), ?biocLite for help
    A newer version of Bioconductor is available for this version of R,
      ?BiocUpgrade for help
    BioC_mirror: https://bioconductor.org
    Using Bioconductor 3.7 (BiocInstaller 1.30.0), R 3.5.1 (2018-07-02).
    Installing package(s) 'Biostrings'
    

    package 'Biostrings' successfully unpacked and MD5 sums checked
    
    The downloaded binary packages are in
    	C:\Users\Administrator\AppData\Local\Temp\RtmpG0ZIpG\downloaded_packages
    

    Old packages: 'ade4', 'backports', 'BH', 'BiocManager', 'broom', 'callr',
      'caret', 'checkpoint', 'class', 'cli', 'clipr', 'codetools', 'colorspace',
      'curl', 'data.table', 'dbplyr', 'ddalpha', 'digest', 'dimRed', 'doParallel',
      'dplyr', 'evaluate', 'fansi', 'foreign', 'ggplot2', 'haven', 'htmlwidgets',
      'httpuv', 'httr', 'ipred', 'IRdisplay', 'IRkernel', 'jsonlite', 'kernlab',
      'knitr', 'later', 'lattice', 'lava', 'magic', 'markdown', 'MASS', 'Matrix',
      'mgcv', 'mime', 'MKmisc', 'ModelMetrics', 'openssl', 'pillar', 'pkgconfig',
      'pls', 'processx', 'purrr', 'R6', 'Rcpp', 'readr', 'readxl', 'recipes',
      'repr', 'reprex', 'rlang', 'rmarkdown', 'robustbase', 'rstudioapi', 'RUnit',
      'scales', 'sfsmisc', 'shiny', 'stringi', 'survival', 'testthat', 'tibble',
      'tidyr', 'tidyselect', 'tinytex', 'TTR', 'xfun', 'xtable', 'xts', 'zoo'
    Loading required package: BiocGenerics
    Loading required package: parallel
    
    Attaching package: 'BiocGenerics'
    
    The following objects are masked from 'package:parallel':
    
        clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
        clusterExport, clusterMap, parApply, parCapply, parLapply,
        parLapplyLB, parRapply, parSapply, parSapplyLB
    
    The following objects are masked from 'package:stats':
    
        IQR, mad, sd, var, xtabs
    
    The following objects are masked from 'package:base':
    
        anyDuplicated, append, as.data.frame, basename, cbind, colMeans,
        colnames, colSums, dirname, do.call, duplicated, eval, evalq,
        Filter, Find, get, grep, grepl, intersect, is.unsorted, lapply,
        lengths, Map, mapply, match, mget, order, paste, pmax, pmax.int,
        pmin, pmin.int, Position, rank, rbind, Reduce, rowMeans, rownames,
        rowSums, sapply, setdiff, sort, table, tapply, union, unique,
        unsplit, which, which.max, which.min
    
    Loading required package: S4Vectors
    Loading required package: stats4
    
    Attaching package: 'S4Vectors'
    
    The following object is masked from 'package:base':
    
        expand.grid
    
    Loading required package: IRanges
    
    Attaching package: 'IRanges'
    
    The following object is masked from 'package:grDevices':
    
        windows
    
    Loading required package: XVector
    
    Attaching package: 'Biostrings'
    
    The following object is masked from 'package:base':
    
        strsplit
    
    

2. Read the two sequences. In the following example, we have shown manually
typed-in sequences for convenience, but the same method can be used for
other types of sequences as well:


```R
sequence1 <- "GAATTCGGCTA"
sequence2 <- "GATTACCTA"
```

3. Assign penalties for the mismatch and gap to get a scoring matrix for the nucleotides
as follows:


```R
myScoringMat <- nucleotideSubstitutionMatrix(match = 1, mismatch= -1, baseOnly = TRUE)
myScoringMat
```


<table>
<thead><tr><th></th><th scope=col>A</th><th scope=col>C</th><th scope=col>G</th><th scope=col>T</th></tr></thead>
<tbody>
	<tr><th scope=row>A</th><td> 1</td><td>-1</td><td>-1</td><td>-1</td></tr>
	<tr><th scope=row>C</th><td>-1</td><td> 1</td><td>-1</td><td>-1</td></tr>
	<tr><th scope=row>G</th><td>-1</td><td>-1</td><td> 1</td><td>-1</td></tr>
	<tr><th scope=row>T</th><td>-1</td><td>-1</td><td>-1</td><td> 1</td></tr>
</tbody>
</table>



4. Assign gap penalties for the alignments as follows:


```R
gapOpen <- 2
gapExtend <- 1
```

5. Run the following pairwiseAlignment function to perform a global alignment for
the sequences:


```R
myAlignment <- pairwiseAlignment(sequence1, sequence2,substitutionMatrix = myScoringMat, gapOpening = gapOpen,gapExtension = gapExtend, type="global", scoreOnly = FALSE)
myAlignment
```


    Global PairwiseAlignmentsSingleSubject (1 of 1)
    pattern: GAATTCGGCTA
    subject: GATTAC--CTA
    score: 1 


6. In the case of protein sequences, we have scoring matrices that are called
substitution matrices. To see them, use the following function:


```R
data(package="Biostrings")
```

7. Assign/select one of these substitution matrices to align the protein sequences as follows:


```R
data(BLOSUM62)
subMat <- "BLOSUM62"
```

8. Now, perform the alignment of the protein sequences with the selected substitution
matrix as follows:


```R
sequence1 <- "PAWHEAE"
sequence2 <- "HEAGAWGHE"
myAlignProt <- pairwiseAlignment(sequence1, sequence2,substitutionMatrix = subMat, gapOpening = gapOpen, gapExtension =gapExtend, type="global", scoreOnly = FALSE)
```
