
1. Looking for the start and stop codons in R can be seen as a simple pattern-matching
problem and can be solved with the Biostrings library. Load this library with the
following function:


```R
library(Biostrings)
```

    Loading required package: BiocGenerics
    Loading required package: parallel
    
    Attaching package: 'BiocGenerics'
    
    The following objects are masked from 'package:parallel':
    
        clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
        clusterExport, clusterMap, parApply, parCapply, parLapply,
        parLapplyLB, parRapply, parSapply, parSapplyLB
    
    The following objects are masked from 'package:stats':
    
        IQR, mad, sd, var, xtabs
    
    The following objects are masked from 'package:base':
    
        anyDuplicated, append, as.data.frame, basename, cbind, colMeans,
        colnames, colSums, dirname, do.call, duplicated, eval, evalq,
        Filter, Find, get, grep, grepl, intersect, is.unsorted, lapply,
        lengths, Map, mapply, match, mget, order, paste, pmax, pmax.int,
        pmin, pmin.int, Position, rank, rbind, Reduce, rowMeans, rownames,
        rowSums, sapply, setdiff, sort, table, tapply, union, unique,
        unsplit, which, which.max, which.min
    
    Loading required package: S4Vectors
    Loading required package: stats4
    
    Attaching package: 'S4Vectors'
    
    The following object is masked from 'package:base':
    
        expand.grid
    
    Loading required package: IRanges
    
    Attaching package: 'IRanges'
    
    The following object is masked from 'package:grDevices':
    
        windows
    
    Loading required package: XVector
    
    Attaching package: 'Biostrings'
    
    The following object is masked from 'package:base':
    
        strsplit
    
    

2. Then, create the sequence to be analyzed as follows (in our case, we use a random
sequence, but we can use the sequences fetched from GenBank as well):


```R
mynucleotide <- DNAString("aacataatgcagtagaacccatgagccc")
```

3. Look for a pattern of your interest, such as a start codon ATG, as shown in the
following example:


```R
matchPattern(DNAString("ATG"), mynucleotide)
```


      Views on a 28-letter DNAString subject
    subject: AACATAATGCAGTAGAACCCATGAGCCC
    views:
        start end width
    [1]     7   9     3 [ATG]
    [2]    21  23     3 [ATG]


4. Similarly, look for the pattern for the stop codons, such as TAA or other stop codons,
as shown in the following example:


```R
matchPattern("TAA", mynucleotide)
```


      Views on a 28-letter DNAString subject
    subject: AACATAATGCAGTAGAACCCATGAGCCC
    views:
        start end width
    [1]     5   7     3 [TAA]


5. Now, combine these two aspects into a single function to return the overall results for
all the codons as in the following function:


```R
myCodonFinder <- function(sequence){
    startCodon = DNAString("ATG") # Assign start codons
    stopCodons = list("TAA", "TAG", "TGA") # Assign stop codons
    codonPosition = list() #initialize the output to be returned as a list
    codonPosition$Start = matchPattern(startCodon, sequence) #search start codons
    x=list()
for(i in 1:3){   # iterate over all stop codons
    x[[i]]= matchPattern(DNAString (stopCodons[[i]]), sequence)
    codonPosition$Stop=x
}
return(codonPosition) # returns results
}
```

6. Now, paste the code for the previous function into the R session and run it with your
sequence object mynucleotide as follows:


```R
StartStops <- myCodonFinder(mynucleotide)
```

Alternatively, save the source code as a file, say, myCodonFinder.R, and source it
into the R session as followsï¼š

7. To find the genes in the nucleotide sequence, tweak the function by looking for a start
codon and then a potential stop codon with a set of codons in between.

The modification can also be used to find open reading frames (ORFs).
